codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java

```diff
@@ -106,6 +106,9 @@ public long errorCode() {
     private final TreeMap<Integer, PendingStream> pendingStreams = new TreeMap<Integer, PendingStream>();
     private int maxConcurrentStreams;
     private boolean closed;
+    private Integer goAwayLastStreamId;
+    private long goAwayErrorCode;
+    private ByteBuf goAwayDebugData;
```

Comment:
This buffer needs to be released

Reply:
This should be the byte[] as produced from ByteBufUtil.getBytes().

Reply:
Either the type is incorrect then or the release is missing . Also I think as this method may return the underlying byte[] it is not save to store it here. You need to make a copy otherwise it may be reused / released already when used

Reply:
I was agreeing with you. The type should become byte[] and us not worry about its lifetime. I think this is actually broken for the new code path, as a reference isn't donated by onGoAwayReceived() and so this ByteBuf may already be released by the time the new code path runs. The test doesn't notice because it is using EMPTY_BUFFER for debug data.

> You need to make a copy otherwise it may be reused / released already when used

I think that's a general problem with Http2GoAwayException. The same exception is used for many streams. If any of them modify it all of them will see the change.

Reply:
I think we need to do two things ... clone in the constructor + ensure we also store a copy / clone here

Reply:
> clone in the constructor

I don't see how that helps this PR, as a single exception can be received by many pieces of code. It seems the getter `Http2GoAwayException.debugData()` would need to clone. The constructor could defend itself with another clone to handle user code that creates `Http2GoAwayException` (which is mostly theoretical), but that wouldn't impact any of the code being changed in this PR.

Reply:
Now using byte array with clone. Need it be nullified when `close()` is called?

Reply:
I don't think there's any need to null it out.